feat: juju switch to previous controller or model

gfouillet opened the pull request:
It implements a new behavior to `juju switch`, similar to well known shell command `cd -`

It also introduce a flag-based way to switch, through `-c` and `-m` flags (as done in other command):
* `juju switch -m a-model` equivalent to `juju switch :a-model`
* `juju switch -c a-controller` equivalent to `juju switch a-controller:`

To implements the behavior similar with `cd -`, this PR introduce the placeholder `-` which mean "previous". So:

* `juju switch -` # switch to previous state controller:model (current model of previous controller if the controller has changed, previous model of current controller if model has changed)
* `juju switch -m -` # switch to previous model, current controller
* `juju switch -c -` # switch to previous controller

I also added two related but independent commit:

* A commit to remove a duplicate TU for the switch command
* A commit which remove several golang warning from the files I have modified.

## Checklist

<!-- If an item is not applicable, use `~strikethrough~`. -->

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- ~[ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- ~[ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

Unit test: `TEST_PACKAGES="./cmd/juju/commands/... -count=1 -race -check.vv" make run-go-tests`

However jujucli state is stored in a file which use handled by [jujuclient/file.go] which has no UT (and the store implemented there is replace by a in-memory store and thus uncovered). It requires few manual testing.

```sh
# providing we have 
# ctrl1 with m1 & m2
# ctrl2 with m1 & m2

# Arrange
juju switch ctrl1:m1 # set previous model for ctrl1
juju switch m2  # set current model for ctrl1
juju switch ctrl2:m1  # set previous model for ctrl2
juju switch m2  # set current model for ctrl2

# Assert
juju switch -c - # ctrl2:m2 -> ctrl1:m2 (switch controller)
juju switch -m - # ctrl1:m2 -> ctrl1:m1 (switch model)
juju switch - # ctrl1:m1 -> ctrl1:m2 (switch model)
juju switch ctrl2:m1 # ctrl1:m2 -> ctrl2:m1 (switch controller & model)
juju switch - # ctrl2:m1 -> ctrl1:m2 (switch controller)
juju add-model m3 # ctrl1:m2 -> ctrl1:m3 (switch model)
juju switch - # ctrl1:m3 -> ctrl1:m2 (switch model)
```

## Documentation changes

[juju switch](https://juju.is/docs/juju/juju-switch) command documentation need to be updated 

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/1988426

**Jira card:** [JUJU-6168](https://warthogs.atlassian.net/browse/JUJU-6168)

[JUJU-6168]: https://warthogs.atlassian.net/browse/JUJU-6168?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

nvinuesa commented on cmd/juju/commands/switch.go:
```diff
@@ -5,10 +5,10 @@ package commands
 	"github.com/juju/cmd/v4"
 	"github.com/juju/errors"
+	"github.com/juju/gnuflag"
+	"os"
```
The `os` package should go in the first grouping of imports (std lib).

gfouillet replied:
Sure I didn't gci files. Fixed.

nvinuesa commented on cmd/juju/commands/switch_test.go:
```diff
@@ -373,3 +366,130 @@ func (s *SwitchSimpleSuite) TestSwitchCurrentModelNoLongerInStore(c *gc.C) {
+	c.Assert(cmdtesting.Stderr(context), gc.Equals, "currentCtrl (controller) -> previousCtrl (controller)\n")
+}
+
+func (s *SwitchSimpleSuite) TestSwitchPreviousControllerWhichDoesntExists(c *gc.C) {
```
Could you please add a test were you don't have any controller set in `PreviousControllerName`?

nvinuesa commented on cmd/juju/commands/switch_test.go:
```diff
@@ -321,14 +312,16 @@ func (s *SwitchSimpleSuite) TestSwitchUnknownCurrentControllerRefreshModelsFails
 func (s *SwitchSimpleSuite) TestSettingWhenModelEnvVarSet(c *gc.C) {
-	os.Setenv("JUJU_MODEL", "using-model")
-	_, err := s.run(c, "erewhemos-2")
+	err := os.Setenv("JUJU_MODEL", "using-model")
```
Just a thought, should we also accept the previous model and controller as env vars?

nvinuesa reviewed (commented):
First quick pass, just some small comments for the moment.

What should happen in this case? Is this behavior correct? (the setup is just two controllers (`c` and `cc`) with no models on any of them):
```
juju switch -:-
cc (controller) -> c (controller)
juju switch :-
ERROR previous model for controller c not found
```

nvinuesa commented:
You will also need to regenerate some mocks (see the `Generate` check job):
```
go generate ./cmd/juju/commands/...
go generate ./cmd/juju/user/...
```

hpidcock commented:
@gfouillet this should target 3.6 branch.

jack-w-shaw commented on cmd/juju/commands/switch.go:
```diff
@@ -74,9 +80,28 @@ func (c *switchCommand) Info() *cmd.Info {
+func (c *switchCommand) SetFlags(f *gnuflag.FlagSet) {
+	c.CommandBase.SetFlags(f)
+	// The `-:` requires a special handling because it will be detected as a flag by command library
+	f.BoolVar(&c.IsPreviousController, ":", false, "Switch to previous controller")
```
:vomiting_face:

jack-w-shaw replied:
This results in:
```
Command Options:
-:  (= false)
    Switch to previous controller
```
in help text. Is there a way this can be suppressed?

gfouillet replied:
Yes... I know. And agree :D

I can try to dig into the cmd libraries (I know how to do this kind of magic in `spf13/cobra` but in the lib we use... I will dig into it.

gfouillet replied:
There is no way to prevent it, unfortunately. @jack-w-shaw 

To avoid this ugliness we can:

- use another character than `-` as placeholder for previous (but it would make it less intuitive)
- update https://github.com/juju/gnuflag to have "hidden flag" (flag that doesn't show up in automatec documentation)
- update https://github.com/juju/cmd to be able to catch up argument before flag parsing and be able to do some smuggle with args before actually parse argument (or be able to deactivate arg parsing)

jack-w-shaw commented on cmd/juju/commands/switch.go:
```diff
@@ -74,9 +80,28 @@ func (c *switchCommand) Info() *cmd.Info {
+		return err
+	}
+	// handling special arguments like -:, -, and :- is done as part of initialization to outline their
+	//"syntaxic sugar" kind
```
Missing space at between `//` and start of comment on this line

jack-w-shaw commented on cmd/juju/commands/switch.go:
```diff
@@ -47,7 +50,7 @@ default model in mycontroller, :mymodel switches to mymodel in current
 The `[1:] + "`juju models`" + ` command can be used to determine the active model
 (of any controller). An asterisk denotes it.
-
+The special arguments - (hyphen) allows to return to previous model or controller.
```
```
Details:
When used without an argument, the command shows the current controller
and its active model.
When a single argument without a colon is provided juju first looks for a
controller by that name and switches to it, and if it's not found it tries
to switch to a model within current controller. mycontroller: switches to
default model in mycontroller, :mymodel switches to mymodel in current
controller and mycontroller:mymodel switches to mymodel on mycontroller.
The `juju models` command can be used to determine the active model
(of any controller). An asterisk denotes it.
The special arguments - (hyphen) allows to return to previous model or controller.
```

This doesn't look very nice. While you're here, could you add some white space?

gfouillet replied:
Yes I can. Maybe i can refresh the whole text likewise:
@jack-w-shaw 

```sh
Details:
When used without an argument, the command shows the current controller
and its active model.

When a single argument without a colon is provided juju first looks for a
controller by that name and switches to it, and if it's not found it tries
to switch to a model within current controller. 

Colon allows to disambiguate model over controller:
- mycontroller: switches to default model in mycontroller, 
- :mymodel switches to mymodel in current controller 
- mycontroller:mymodel switches to mymodel on mycontroller.

The special arguments - (hyphen) instead of a model or a controller allows to return 
to previous model or controller.

The `juju models` command can be used to determine the active model
(of any controller). An asterisk denotes it.

```

jack-w-shaw commented:
I like `-:` and `:-`, but I'm not too keen on the behaviour of `-` by itself.

Swapping to the previous controller if there are multiple, previous model if there is only 1 is a good compromise, but a compromise nonetheless. Ideally I would like to to swap to the previous model, irregardless of what controller we were on.

This current behaviour leads to some weirdness. Say we have:
```
$ juju controllers
Use --refresh option with this command to see the latest information.

Controller  Model  User   Access     Cloud/Region         Models  Nodes    HA  Version
lxd*        m      admin  superuser  localhost/localhost       3      1  none  3.4.4.2      
lxd2        m      admin  superuser  localhost/localhost       2      1  none  3.6-beta1.1  

$ juju models -c lxd
Controller: lxd

Model       Cloud/Region         Type  Status     Machines  Units  Access  Last connection
controller  localhost/localhost  lxd   available         1      1  admin   just now
m*          localhost/localhost  lxd   available         2      1  admin   29 seconds ago

$ juju models -c lxd2
Controller: lxd2

Model       Cloud/Region         Type  Status     Machines  Units  Access  Last connection
controller  localhost/localhost  lxd   available         1      1  admin   just now
m2          localhost/localhost  lxd   available         0      -  admin   10 seconds ago
m3          localhost/localhost  lxd   available         0      -  admin   10 seconds ago
```

Then run: 
```
$ juju switch lxd2:m2
lxd:admin/m -> lxd2:admin/m2

$ juju switch lxd2:m3 
lxd2:admin/m2 -> lxd2:admin/m3
```

I would want `juju switch -` to switch to `lxd2:m2`, however we switch to `lxd:m`.

I understand this would be a little fiddly though. All that would be needed is a boolean which designates if the previous switch was inter or intra controller, which you would use in `resolvePreviousTarget` to branch between the two cases. There isn't a natural place to store this though, it feels out of place in `controllers.yaml` to me, but would (in my opinion) significantly improve this feature

NOTE: When you do an annonymous switch to the previous controller, you default to the 'current model' for that controller, which is exactly what we want. So `-` works perfectly if your last switch was across controllers. The only issue is if we switch within a controller

gfouillet commented:
@jack-w-shaw  I think it's a good suggestion (https://github.com/juju/juju/pull/17516#issuecomment-2173226408)
 
I don't think it's that complicated to implements. In fact there is one complicated way and one simple way with possible yet probably limited side effect:

 ### Simple with side effect:

Just ALWAYS Set current controller even if hasn't changed. I will update the previous one 

As a side effect is more io (status file will always been writed twice), but we are in a cli command which shouldn't be called often.

### Complicated

Don't hide previous maintenance code in  SetCurrentXXX method of the store, but delegate everything to switch.go. I don't think it's a good idea, because it make the code a bit more complicated. However it may be discussed, because above solution may be seen as relying on side effect of SetCurrentController.

gfouillet commented:
@jack-w-shaw I think it's a good suggestion. I will dig into find how to implement this behavior and figure out some issue which would make it counter intuitive.

gfouillet commented:
@jack-w-shaw I updated documentation and change the behavior of `juju switch -` as requested.

I also completely refactor the  code to be able to do it :D

jack-w-shaw commented:
/build

jack-w-shaw reviewed (changes requested):
We're still not quite there

```
$ juju bootstrap lxd lxd1
$ juju bootstrap lxd lxd2
$ juju add-model m
Added 'm' model on localhost/localhost with credential 'localhost' for user 'admin'
$ juju switch lxd1
lxd2:admin/m -> lxd1 (controller)
$ juju add-model m
Added 'm' model on localhost/localhost with credential 'localhost' for user 'admin'
$ juju add-model n
Added 'n' model on localhost/localhost with credential 'localhost' for user 'admin'
$ juju switch m
lxd1:admin/n -> lxd1:admin/m
$ juju switch -
lxd1:admin/m -> lxd2:admin/m
```

I would expect this final switch:
```
$ juju switch -
lxd1:admin/m -> lxd1:admin/n
```

jack-w-shaw commented on cmd/juju/commands/switch.go:
```diff
@@ -74,9 +91,34 @@ func (c *switchCommand) Info() *cmd.Info {
+	if c.modelName != "" || c.controllerName != "" {
+		// translate `-c ctrl` to `ctrl:` and `-m model` to `:model`
+		target = fmt.Sprintf("%s:%s", c.controllerName, c.modelName)
+		err = cmd.CheckEmpty(args)
```
This doesn't feel nice. Do we really need to join the model and controller name? Later down the line we split them, I think we should keep them separate here.

We could even move the (potential) split of c.Target into Init here

I also don't think we need to support `:-` and `-:` anymore (in fact I think `-:` would error now because it's not a flag?). Can we bypass this entirely?

jack-w-shaw replied:
This is turning out to be very fiddly :sweat_smile:

gfouillet replied:
It's because amho, the `-` is more of a trick. Main code support `controller:`, `:model`, `controller-or-model` or `controller:model`. So it make sens to use those format as go-for-all. But I understand your feeling, even if I am not sure that we may do better. The fact that we need to handle `controller-or-model` usecase make separating controller and model a tricky case. But maybe we can merge both and have a function that retun `(controller, model string, err error)` and just set Controller & model in Run. however I am not sure that the code would be more readable.

Could need a second pair of eyes when doing that stuff... Are you open to peer prog on monday ?

jack-w-shaw replied:
> It's because amho, the `-` is more of a trick. Main code support `controller:`, `:model`, `controller-or-model` or `controller:model`. So it make sens to use those format as go-for-all. But I understand your feeling, even if I am not sure that we may do better. The fact that we need to handle `controller-or-model` use case make separating controller and model a tricky case. But maybe we can merge both and have a function that retun `(controller, model string, err error)` and just set Controller & model in Run. however I am not sure that the code would be more readable.

Yeah the `controller-or-model` case does complicate things a little. We may need a special case for this indeterminate form.

Other than that though, the forms I see are all determinate, those being (laying out for sake of my own thoughts):
- `juju switch $controller:$model`
- `juju switch :$model`
- `juju switch $controller:`
- `juju switch -c $controller -m $model`
- `juju switch -c $controller`
- `juju switch -m $model`
Where `$controller` and `$model` can be `-`.

(args and flags should be mutually exclusive)

First 3 cases, I think in `Init` we should parse by splitting on `:` to determine the target controller & model which are then loaded to the `switchCommand` struct and read in `Run`

Final 3 cases, we can load straight into the struct using `SetFlags`

So if an arg is provided and doesn't contain a `:`, maybe load that into a separate attribute and branch in `Run` if it's set?

> Could need a second pair of eyes when doing that stuff... Are you open to peer prog on monday ?

I'm away until Wednesday next week, but yes I will be happy to then

gfouillet replied:
I like your suggestion, but I cannot decide in my head if it would be easier to read.

We can use the pair programming session on Wednesday, it's not an urgent task. Enjoy your time off ;)

gfouillet commented:
> We're still not quite there
> 
> ```
> $ juju bootstrap lxd lxd1
> $ juju bootstrap lxd lxd2
> $ juju add-model m
> Added 'm' model on localhost/localhost with credential 'localhost' for user 'admin'
> $ juju switch lxd1
> lxd2:admin/m -> lxd1 (controller)
> $ juju add-model m
> Added 'm' model on localhost/localhost with credential 'localhost' for user 'admin'
> $ juju add-model n
> Added 'n' model on localhost/localhost with credential 'localhost' for user 'admin'
> $ juju switch m
> lxd1:admin/n -> lxd1:admin/m
> $ $ juju switch -
> lxd1:admin/m -> lxd2:admin/m
> ```
> 
> I would expect this final switch:
> 
> ```
> $ juju switch -
> lxd1:admin/m -> lxd1:admin/n
> ```

Yes me too...

```
$ juju bootstrap lxd lxd1 
# curctrl = lxd1, prevCtrl=none, no model on lxd1, ctrl-changed-on-prv-switch=true
$ juju bootstrap lxd lxd2 
# curctrl = lxd2, prevCtrl=lxd1, no model on lxd1 nor lxd2, ctrl-changed-on-prv-switch=true
$ juju add-model m  
# curctrl = lxd2, prevCtrl=lxd1, curModel(lxd2)=m, prevModel(lx2)=none, no model on lxd1, ctrl-changed-on-prv-switch=true <=== should false (we switched model... technically)
$ juju switch lxd1  
# curctrl = lxd1, prevCtrl=lxd2, curModel(lxd2)=m, prevModel(lxd2)=none, no model on lxd1, ctrl-changed-on-prv-switch=true 
$ juju add-model m  
# curctrl = lxd1, prevCtrl=lxd2, curModel(lxd2)=m, prevModel(lxd2)=none, curModel(lxd1)=m, prevModel(lxd1)=none, ctrl-changed-on-prv-switch=true  <=== should false (we switched model... technically)
$ juju add-model n
# curctrl = lxd1, prevCtrl=lxd2, curModel(lxd2)=m, prevModel(lxd2)=none, curModel(lxd1)=n, prevModel(lxd1)=m, nonectrl-changed-on-prv-switch=true  <=== should false (we switched model... technically)
$ juju switch m
# curctrl = lxd1, prevCtrl=lxd2, curModel(lxd2)=m, prevModel(lxd2)=none, curModel(lxd1)=m, prevModel(lxd1)=n, nonectrl-changed-on-prv-switch=false
$ juju switch -
 lxd1:admin/m -> lxd1:admin/n # works to me
 ```
 
 I think you found a bug but your use scenario is wrong.
 
However the minimal use case I found is:

```sh
$ juju bootstrap lxd lxd1
$ juju bootstrap lxd lxd2
$ juju add-model m
Added 'm' model on localhost/localhost with credential 'localhost' for user 'admin'
$ juju switch -
lxd2:admin/m -> lxd1 (controller)
```
And should be:
```sh
$ juju switch -
ERROR previous model for controller lxd2 not found
```

Second use case:

```sh
$ juju bootstrap lxd lxd1
$ juju bootstrap lxd lxd2
$ juju add-model m
Added 'n' model on localhost/localhost with credential 'localhost' for user 'admin'
$ juju add-model n
Added 'm' model on localhost/localhost with credential 'localhost' for user 'admin'
$ juju switch -
lxd2:admin/n -> lxd1 (controller)
```

should be:
```sh
$ juju switch -
lxd2:admin/n -> lxd2:admin/m
```

@jack-w-shaw Do you agree ? If yes, I need to find a way to explicitly switch controller/model when adding model.

jack-w-shaw commented:
> > We're still not quite there
> > ```
> > $ juju bootstrap lxd lxd1
> > $ juju bootstrap lxd lxd2
> > $ juju add-model m
> > Added 'm' model on localhost/localhost with credential 'localhost' for user 'admin'
> > $ juju switch lxd1
> > lxd2:admin/m -> lxd1 (controller)
> > $ juju add-model m
> > Added 'm' model on localhost/localhost with credential 'localhost' for user 'admin'
> > $ juju add-model n
> > Added 'n' model on localhost/localhost with credential 'localhost' for user 'admin'
> > $ juju switch m
> > lxd1:admin/n -> lxd1:admin/m
> > $ $ juju switch -
> > lxd1:admin/m -> lxd2:admin/m
> > ```
> > 
> > I would expect this final switch:
> > ```
> > $ juju switch -
> > lxd1:admin/m -> lxd1:admin/n
> > ```
> 
> Yes me too...
> 
> ```
> $ juju bootstrap lxd lxd1 
> # curctrl = lxd1, prevCtrl=none, no model on lxd1, ctrl-changed-on-prv-switch=true
> $ juju bootstrap lxd lxd2 
> # curctrl = lxd2, prevCtrl=lxd1, no model on lxd1 nor lxd2, ctrl-changed-on-prv-switch=true
> $ juju add-model m  
> # curctrl = lxd2, prevCtrl=lxd1, curModel(lxd2)=m, prevModel(lx2)=none, no model on lxd1, ctrl-changed-on-prv-switch=true <=== should false (we switched model... technically)
> $ juju switch lxd1  
> # curctrl = lxd1, prevCtrl=lxd2, curModel(lxd2)=m, prevModel(lxd2)=none, no model on lxd1, ctrl-changed-on-prv-switch=true 
> $ juju add-model m  
> # curctrl = lxd1, prevCtrl=lxd2, curModel(lxd2)=m, prevModel(lxd2)=none, curModel(lxd1)=m, prevModel(lxd1)=none, ctrl-changed-on-prv-switch=true  <=== should false (we switched model... technically)
> $ juju add-model n
> # curctrl = lxd1, prevCtrl=lxd2, curModel(lxd2)=m, prevModel(lxd2)=none, curModel(lxd1)=n, prevModel(lxd1)=m, nonectrl-changed-on-prv-switch=true  <=== should false (we switched model... technically)
> $ juju switch m
> # curctrl = lxd1, prevCtrl=lxd2, curModel(lxd2)=m, prevModel(lxd2)=none, curModel(lxd1)=m, prevModel(lxd1)=n, nonectrl-changed-on-prv-switch=false
> $ juju switch -
>  lxd1:admin/m -> lxd1:admin/n # works to me
> ```

Huh, weird. Let me double check.

> I think you found a bug but your use scenario is wrong.
> 
> However the minimal use case I found is:
> 
> ```shell
> $ juju bootstrap lxd lxd1
> $ juju bootstrap lxd lxd2
> $ juju add-model m
> Added 'm' model on localhost/localhost with credential 'localhost' for user 'admin'
> $ juju switch -
> lxd2:admin/m -> lxd1 (controller)
> ```
> 
> And should be:
> 
> ```shell
> $ juju switch -
> ERROR previous model for controller lxd2 not found
> ```
> 
> Second use case:
> 
> ```shell
> $ juju bootstrap lxd lxd1
> $ juju bootstrap lxd lxd2
> $ juju add-model m
> Added 'n' model on localhost/localhost with credential 'localhost' for user 'admin'
> $ juju add-model n
> Added 'm' model on localhost/localhost with credential 'localhost' for user 'admin'
> $ juju switch -
> lxd2:admin/n -> lxd1 (controller)
> ```
> 
> should be:
> 
> ```shell
> $ juju switch -
> lxd2:admin/n -> lxd2:admin/m
> ```
> 
> @jack-w-shaw Do you agree ? If yes, I need to find a way to explicitly switch controller/model when adding model.

This is something different to what I observed (pretty sure), but is still a valid bug, but agree with your expectation. The solution should be fairly easy, `add-model` should also just update the `~/.local/share/juju` config as well right?

gfouillet commented:
Yep, I have the fix on my laptop. Will commit it on Monday.

jack-w-shaw commented on jujuclient/file.go:
```diff
@@ -266,10 +285,10 @@ func (s *store) SetCurrentController(name string) error {
 	}
-	if controllers.CurrentController == name {
-		return nil
+	if controllers.HasControllerChangedOnPreviousSwitch = controllers.CurrentController != name; controllers.HasControllerChangedOnPreviousSwitch {
```
I think this is trying to be too clever. I would rather you explicitly set HasControllerChangedOnPreviousSwitch

jack-w-shaw commented:
> > > We're still not quite there
> > > ```
> > > $ juju bootstrap lxd lxd1
> > > $ juju bootstrap lxd lxd2
> > > $ juju add-model m
> > > Added 'm' model on localhost/localhost with credential 'localhost' for user 'admin'
> > > $ juju switch lxd1
> > > lxd2:admin/m -> lxd1 (controller)
> > > $ juju add-model m
> > > Added 'm' model on localhost/localhost with credential 'localhost' for user 'admin'
> > > $ juju add-model n
> > > Added 'n' model on localhost/localhost with credential 'localhost' for user 'admin'
> > > $ juju switch m
> > > lxd1:admin/n -> lxd1:admin/m
> > > $ $ juju switch -
> > > lxd1:admin/m -> lxd2:admin/m
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I would expect this final switch:
> > > ```
> > > $ juju switch -
> > > lxd1:admin/m -> lxd1:admin/n
> > > ```
> > 
> > 
> > Yes me too...
> > ```
> > $ juju bootstrap lxd lxd1 
> > # curctrl = lxd1, prevCtrl=none, no model on lxd1, ctrl-changed-on-prv-switch=true
> > $ juju bootstrap lxd lxd2 
> > # curctrl = lxd2, prevCtrl=lxd1, no model on lxd1 nor lxd2, ctrl-changed-on-prv-switch=true
> > $ juju add-model m  
> > # curctrl = lxd2, prevCtrl=lxd1, curModel(lxd2)=m, prevModel(lx2)=none, no model on lxd1, ctrl-changed-on-prv-switch=true <=== should false (we switched model... technically)
> > $ juju switch lxd1  
> > # curctrl = lxd1, prevCtrl=lxd2, curModel(lxd2)=m, prevModel(lxd2)=none, no model on lxd1, ctrl-changed-on-prv-switch=true 
> > $ juju add-model m  
> > # curctrl = lxd1, prevCtrl=lxd2, curModel(lxd2)=m, prevModel(lxd2)=none, curModel(lxd1)=m, prevModel(lxd1)=none, ctrl-changed-on-prv-switch=true  <=== should false (we switched model... technically)
> > $ juju add-model n
> > # curctrl = lxd1, prevCtrl=lxd2, curModel(lxd2)=m, prevModel(lxd2)=none, curModel(lxd1)=n, prevModel(lxd1)=m, nonectrl-changed-on-prv-switch=true  <=== should false (we switched model... technically)
> > $ juju switch m
> > # curctrl = lxd1, prevCtrl=lxd2, curModel(lxd2)=m, prevModel(lxd2)=none, curModel(lxd1)=m, prevModel(lxd1)=n, nonectrl-changed-on-prv-switch=false
> > $ juju switch -
> >  lxd1:admin/m -> lxd1:admin/n # works to me
> > ```
> 
> Huh, weird. Let me double check.

I have double checked this, you are right, this behaves as expected :+1:

jack-w-shaw commented:
I think you should refactor along the lines of https://github.com/juju/juju/pull/17643

There are also some unit tests that need fixing

jack-w-shaw commented on cmd/juju/commands/switch.go:
```diff
@@ -232,3 +319,44 @@ func (c *switchCommand) name(store jujuclient.ModelGetter, controllerName string
+func (c *switchCommand) trySwitchToModel(store modelcmd.QualifyingClientStore, controller string, model string) (string, error) {
+	if err := store.SetCurrentController(controller); err != nil {
+		return "", errors.Trace(err)
+	}
```
Why don't we check the controller exists like in `trySwitchToController`?

gfouillet replied:
It is not necessary since 
`modelName, err := store.QualifiedModelName(controller, model)` will fails with a pretty accurate message if the controller doesn't exist:

```sh
❯ juju switch jack:shaw
ERROR controller jack not found
```

jack-w-shaw commented on cmd/juju/commands/switch.go:
```diff
@@ -74,10 +91,104 @@ func (c *switchCommand) Info() *cmd.Info {
+	}
+
+	return nil
+}
```
`errors.Trace` should (almost) all be replaced with `errors.Annotate` or `errors.Errorf`

gfouillet replied:
I will add it on Run function. Seems that annotate in Init are discarded anyway (even with --debug)

jack-w-shaw commented on jujuclient/controllers.go:
```diff
@@ -72,4 +72,10 @@ type Controllers struct {
+	PreviousController string `yaml:"previous-controller,omitempty"`
+
+	// HasControllerChangedOnPreviousSwitch represents whether the controller has changed during the previous switch.
+	HasControllerChangedOnPreviousSwitch bool `yaml:"has-controller-changed-on-previous-switch"`
```
Is there really no way to improve this?

I can't think of any of the top of my head. But something else would be good. Even if it results in redundancy

Maybe `global-previous-model` with full `$controller:$model` format?

gfouillet replied:
I thought about this solution, but i don't like it, because it add "model" information in the controller list context. It doesn't feels right.

Having a boolean which state that the previous switch was extra or intra controller is still bounded to controllers level (whatever the actual content of a controller), so even if it's ugly, I think it's a better choice.

jack-w-shaw commented on jujuclient/interface.go:
```diff
@@ -231,6 +231,12 @@ type ControllerGetter interface {
+	// If there is no previous controller (for instance controller has never been switched),
+	// an error satisfying errors.IsNotFound will be returned.
+	// It also retrieve if the last switch update the current controller or just shift the model
+	PreviousController() (string, bool, error)
```
This signature is also a bad result of `HasControllerChangedOnPreviousSwitch`. This is quite nasty. It isn't at al obvious from the function name

If we can't improve the underlying data representation, I think this `bool` should be returned by it's own function

gfouillet replied:
I don't think it's a good things to add another method to the interface.
I don't see the "previous controller" and "the has Controller changed on previous switch" as independent data. They are coupled (they evolve in response to the same events).

However, if you don't like the return signature, maybe we can inverse the logic:

```go
	// If there is no previous controller (for instance controller has never been switched),
	// an error satisfying errors.IsNotFound will be returned.
	// if force is true, it will always return the previous controller.
	// If force is false, it may return the current controller if a previous switch hasn't changed the current controller.
	PreviousController(force bool) (string, error)
```

jack-w-shaw replied:
I don't like this either really. Quite a conundrum

gfouillet replied:
Actually, thinking about it, this latest signature is worse, because it buries more logic in the store which it shouldn't (moreover, the store is completly mocked in the UT, so this later solution is worse, since the behavior won't be tested)

So, after sleeping on it we have two solution:

- Keep it that way (with a boolean returned by the PreviousController stuff)
- split into two method.

The former put the emphasis on the fact that "Previous controller" and "ControllerHasSwitched" is the same crap.

The later don't define such emphasis (which i think maybe an issue at some point), but may be easier to read.

Honestly, i can go for any tie break from @SimonRichardson, i don't have a strong enough conviction on it.

jack-w-shaw reviewed (commented):
One or two mort thoughts.

Nearly there!

SimonRichardson commented on jujuclient/mem.go:
```diff
@@ -336,6 +354,24 @@ func (c *MemStore) CurrentModel(controller string) (string, error) {
 	return controllerModels.CurrentModel, nil
 }
 
+// PreviousModel implements ModelGetter.
```
We could improve the documentation here.

gfouillet replied:
I followed the style. With a good IDE it's good enough...

[Screencast from 04-07-2024 15:09:10.webm](https://github.com/juju/juju/assets/172264669/6dfb8514-6146-49ae-a806-128fc54b1dd1)


Maybe having a more qualified name would be quicker, but i am not convinced yet.

[Screencast from 04-07-2024 15:10:07.webm](https://github.com/juju/juju/assets/172264669/e0e7fc10-83d9-48a0-ad77-7ae1a1043e52)

What do you suggest ?

SimonRichardson commented on jujuclient/jujuclienttesting/stub.go:
```diff
@@ -204,6 +214,12 @@ func (c *StubStore) CurrentController() (string, error) {
 	return c.CurrentControllerFunc()
 }
 
+// PreviousController implements ControllersGetter.PreviousController.
```
```suggestion
// PreviousController implements ControllersGetter.PreviousController.
```

Not a fan of this comment, copy the comment, it provides better information.

SimonRichardson reviewed (commented):
Short of refactoring the files (controller vs model yaml files), and the dichotomy between the two, I think this is ok for now, yet I agree with @jack-w-shaw. The concept isn't pretty and storing state in the files can be problematic if not done correctly. Luckily, we don't have to provide any backwards compatibility for files in the `local/share/juju` otherwise all of this would have to be optional.

Also, this is non-optimial if you've not got a value stored in the previous model/controller.
(You can get into this position by removing the new fields in the yaml files then calling `juju switch -` - I think suggesting to use `juju switch` without `-` until it's populated).

```
$ juju switch -
ERROR interpreting "-" argument after a model switch: previous model for controller src40 not found
```

SimonRichardson commented:
/build

gfouillet commented:
/build

gfouillet commented:
@SimonRichardson All check pass, except the microk8s one which is not required and is consistently failing on PR targeting 3.6.

SimonRichardson reviewed (commented):
@gfouillet can we retarget this to 4.0, that way we can bake it in. Then if we like it we can always backport it.

gfouillet commented:
@SimonRichardson All check pass, except the microk8s one which is not required and is consistently failing on PR targeting 3.6.

> @gfouillet can we retarget this to 4.0, that way we can bake it in. Then if we like it we can always backport it.

Ok, I don't have any preference on the subject. Rebase on going.

SimonRichardson commented:
/build

gfouillet commented:
/merge

gfouillet commented:
/merge
